src/data/enum.lean

```diff
@@ -0,0 +1,190 @@
+
+import category.monad.basic
+import data.list.basic
+import data.equiv.basic
+import data.finset
+import data.fintype
+
+open finset (hiding singleton)
+
+namespace fin
+
+def enum (n : ℕ) : list (fin n) :=
+(list.range n).attach.map $ λ ⟨i,h⟩, ⟨i,by simpa using h⟩
+
+@[simp] lemma mem_enum {n} (x : fin n) : x ∈ enum n :=
+by simp [enum]; existsi x.1; simp [fin.is_lt]
+
+end fin
+
+class enumerable (α : Sort*) :=
+(card : ℕ)
+(equiv : α ≃ fin card)
+
+namespace enumerable
+
+variables {α : Type*}
+
+instance [enumerable α] : decidable_eq α :=
+equiv.decidable_eq_of_equiv (equiv α)
+
+def of_equiv (α) {β} [enumerable α] (h : β ≃ α) : enumerable β :=
+{ card := card α,
+  equiv := h.trans (equiv α)  }
+
+def of_list [decidable_eq α] (xs : list α) (h : ∀ x : α, x ∈ xs) : enumerable α :=
```

Comment:
Maybe it's nicer to define `of_distinct_list` (or some better name) which takes the additional hypothesis that the members of the list are all distinct, and then define `of_list` in terms of it using `erase_dup`.

Reply:
Would you do that to remove the need for `decidable_eq`? That looks moot to me because `decidable_eq` is derivable from `enumerable` so we end up with `decidable_eq` anyway.

Reply:
No, just so that one has better definitional control over the `equiv`, in case one wants that for some reason.